esp32/machine_uart.c

```diff
@@ -0,0 +1,343 @@
+    }
+
+    // set stop bits
+    switch (args[ARG_stop].u_int) {
+        // FIXME: ESP32 also supports 1.5 stop bits
+        case 0:
+            break;
+        case 1:
+            uart_set_stop_bits(self->uart_num, UART_STOP_BITS_1);
+            self->stop = 1;
+            break;
+        case 2:
+            uart_set_stop_bits(self->uart_num, UART_STOP_BITS_2);
+            self->stop = 2;
+            break;
+        default:
+            nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError, "invalid stop bits"));
+            break;
+    }
+
+    // set timeout
+    self->timeout = args[ARG_timeout].u_int;
+
+    // set timeout_char
+    // make sure it is at least as long as a whole character (13 bits to be safe)
+    self->timeout_char = args[ARG_timeout_char].u_int;
+    uint32_t min_timeout_char = 13000 / self->baudrate + 1;
+    if (self->timeout_char < min_timeout_char) {
+        self->timeout_char = min_timeout_char;
+    }
+}
+
+STATIC mp_obj_t machine_uart_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, MP_OBJ_FUN_ARGS_MAX, true);
+
+    // get uart id
+    mp_int_t uart_num = mp_obj_get_int(args[0]);
+    if (uart_num < 0 || uart_num > UART_NUM_MAX) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError, "UART(%d) does not exist", uart_num));
+    }
+
+    // Attempts to use UART0 from Python has resulted in all sorts of fun errors.
+    // FIXME: UART0 is disabled for now.
+    if (uart_num == UART_NUM_0) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_ValueError, "UART(%d) is disabled (dedicated to REPL)", uart_num));
+    }
+
+     // Defaults
+    uart_config_t uartcfg = {
+        .baud_rate = 115200,
+        .data_bits = UART_DATA_8_BITS,
+        .parity = UART_PARITY_DISABLE,
+        .stop_bits = UART_STOP_BITS_1,
+        .flow_ctrl = UART_HW_FLOWCTRL_DISABLE,
+        .rx_flow_ctrl_thresh = 0
+    };
+
+    // create instance
+    machine_uart_obj_t *self = m_new_obj(machine_uart_obj_t);
+    self->base.type = &machine_uart_type;
+    self->uart_num = uart_num;
+    self->baudrate = 115200;
+    self->bits = 8;
+    self->parity = 0;
+    self->stop = 1;
+    self->rts = UART_PIN_NO_CHANGE;
+    self->cts = UART_PIN_NO_CHANGE;
+    self->timeout = 0;
+    self->timeout_char = 0;
+
+    switch (uart_num) {
+        case UART_NUM_0:
+            self->rx = UART_PIN_NO_CHANGE; // GPIO 3
+            self->tx = UART_PIN_NO_CHANGE; // GPIO 1
+            break;
+        case UART_NUM_1:
+            self->rx = 9;
+            self->tx = 10;
+            break;
+        case UART_NUM_2:
+            self->rx = 16;
+            self->tx = 17;
+            break;
+    }
+
+    // init the peripheral
+    // Setup
+    uart_param_config(self->uart_num, &uartcfg);
+
+    // RX and TX buffers hardcoded at 2048 bytes.
+    uart_driver_install(uart_num, 1024 * 2, 1024 * 2, 10, &UART_QUEUE[self->uart_num], 0);
```

Comment:
RX buffer should be at least `UART_FIFO_LEN` (128), but TX buffer can be 0, and it'll just block while sending.
How about 256B for RX buffer? You get NMEA sentences that are 130+ bytes.
64B TX sounds fine.
I agree that ideally it would be configurable at run time, but not quite sure how that can be implemented right now.